javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,479 @@
+        set("people", people)
+        set("company", SeqType(ScopeType(emptyScope.builder().set("people", people).build())))
+      }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from c in company from p in c.people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_where_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+      set("x", TextType)
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers where p < 10 select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers where 2 + 3 select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("from p in numbers\nwhere 2.0 + 3.0 must evaluate to a boolean type" +
+                        " but was PrimitiveType(Double).")
+    )
+  }
+
+  @Test
+  fun check_select_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select new Foo { x: p, y: p + 1i, z: p == 2 }"),
+      SeqType(
+        ScopeType(
+          emptyScope.builder("Foo").set("x", IntType).set("y", IntType).set("z", BooleanType)
+            .build()
+        )
+      ),
+      scope
+    )
+  }
+
+  @Test
+  fun check_let_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers let x = (p > 5) select x"),
+      SeqType(BooleanType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_orderby_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers orderby p select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers orderby numbers select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("order by expression numbers must be a primitive type" +
+                        " but was SeqType(PrimitiveType(Int)).")
+    )
+  }
+
+  private fun checkFunction(
+    function: String,
+    scope: Expression.Scope,
+    expectedType: InferredType = IntType,
+    arg: String = "numbers",
+    numeric: Boolean = true
+  ) {
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select $function($arg)"),
```

Comment:
Why do you wrap every function invocation in the from/select when the function invocation on its own is a perfectly valid Paxel expression?

Reply:
Just thought it would be good to test the inferencing falling through the qualified expressions.